charmcraft/jujuignore.py

```diff
@@ -0,0 +1,215 @@
+                else:
+                    chunks = []
+                    k = i + 2 if rule[i] == '!' else i + 1
+                    while True:
+                        k = rule.find('-', k, j)
+                        if k < 0:
+                            break
+                        chunks.append(rule[i:k])
+                        i = k + 1
+                        k = k + 3
+                    chunks.append(rule[i:j])
+                    # Escape backslashes and hyphens for set difference (--).
+                    # Hyphens that create ranges shouldn't be escaped.
+                    stuff = '-'.join(s.replace('\\', r'\\').replace('-', r'\-')
+                                     for s in chunks)
+                # Escape set operations (&&, ~~ and ||).
+                stuff = re.sub(r'([&~|])', r'\\\1', stuff)
+                i = j + 1
+                if stuff[0] == '!':
+                    stuff = '^' + stuff[1:]
+                elif stuff[0] in ('^', '['):
+                    stuff = '\\' + stuff
+                res = '%s[%s]' % (res, stuff)
+        elif c == '/':
+            # Special case of '/**/' which can match a single '/'
+            if i < n and rule[i] == '*' and rule[i - 1:i + 3] == '/**/':
+                i = i + 3
+                res = res + '.*/'
+            else:
+                res = res + '/'
+        else:
+            res = res + re.escape(c)
+    return r'(?s:%s)\Z' % res
+
+
+class _Matcher:
+    """Couple a regex with other metadata for how we should match a given pattern."""
+
+    def __init__(self, line_num: int, orig_rule: str, invert: bool, only_dirs: bool,
+                 regex: typing.Pattern):
+        self.line_num = line_num
+        self.orig_rule = orig_rule
+        self.invert = invert
+        self.only_dirs = only_dirs
+        self.regex = regex
+        self.compiled = re.compile(regex)
+
+    def match(self, path: str, is_dir: bool) -> str:
+        """Checks if a path matches.
+
+        Returns:
+            Can return one of KEEP, SKIP, FORCEKEEP
+        """
+        if self.only_dirs and not is_dir:
+            return KEEP
+        if self.compiled.match(path):
+            if self.invert:
+                return FORCEKEEP
+            return SKIP
+        return KEEP
+
+
+class JujuIgnore:
+    """Track a set of ignore patterns from a .jujuignore file."""
+
+    def __init__(self, patterns: typing.Iterable[str]):
+        self._matchers = []
+        self._compile_from(patterns)
+
+    def _compile_from(self, patterns):
+        for line_num, rule in enumerate(patterns):
+            # humans like line numbers to start from 1
+            line_num = line_num + 1
+            orig_rule = rule
+            rule = rule.lstrip()
+            if not rule or rule.startswith('#'):
+                continue
+            invert = False
+            if rule.startswith('!'):
+                invert = True
+                rule = rule[1:]
```

Comment:
AFAIR if the rule starts with multiple bangs juju treats it as a single one, so maybe what we want here is `rule = rule.lstrip('!')`?

Reply:
yeah, that is what juju itself is doing: https://github.com/juju/charm/blob/master/jujuignore.go#L176
I'll add a test